protocol/x/perpetuals/keeper/perpetual.go

```diff
@@ -1372,10 +1370,26 @@ func (k Keeper) GetParams(ctx sdk.Context) types.Params {
 	}
 }
 
-// `SetParams` sets all perpetuals module parameters in store.
-func (k Keeper) SetParams(ctx sdk.Context) error {
-	// TODO(CORE-560): implement this method
-	return status.Errorf(codes.Unimplemented, "SetParams not implemented")
+// `SetParams` atomically sets all perpetuals module parameters in store.
+func (k Keeper) SetParams(ctx sdk.Context, params types.Params) error {
+	// Validate params.
+	if err := params.Validate(); err != nil {
+		return err
+	}
+	// Try to set params in cached store.
+	cacheCtx, writeCache := ctx.CacheContext()
```

Comment:
Why is this and below `writeCache()` needed here? Message handlers should not be responsible for using a branched context

Reply:
I was thinking that `Params` update should succeed atomically for all three parameters. For example, without the branched context, `FundingRateClampFactorPpm` can be updated and then `PremiumVoteClampFactorPpm` update can fail. which I feel like is not the best behavior for `UpdateParams` message

In the future when we migrate `Params` to use just one key, a branched ctx is no longer needed.

Reply:
It's not the responsibility of a message handler to maintain atomicity - in this case this would be handled in `x/gov` endblocker where it uses a cached context, and only writes to cache if the message handler succeeds

Reply:
Ah that makes sense. Will update the PR!